### ISSUE TITLE
v2.0.0-pre33: Specter stops running after long hibernation

### ISSUE THREAD

**Issue by moritzwietersheim** [cryptoadvance/specter-desktop#2278]
- As @k9ert and me discussed yesterday, Specter seems to stop running after long hibernation.

- Please note that this is not an issue that should keep us from releasing v2.0.0 asap.

- Apparently the logs do not provide much information as of now - see attached file for full logs.
- 
- Moritz will run it overnight to see if is still happening on pre34. 

Probably the most relevant lines are this:
```
2023-03-01T04:33:47.365Z [info] : stderr-SPECTERD: INFO in wallet_manager: Updating wallet manager done in  4277ms. Result:
SPECTERD: INFO in wallet_manager:   * loaded_wallets: 2
SPECTERD: INFO in wallet_manager:   * failed_load_wallets: 0

2023-03-01T04:57:20.805Z [error] : specterd exited with code null
2023-03-01T04:57:20.807Z [info] : Updated LoaderMsg: Specter encountered an error:<br>Specter exited with exit code null. Check the logs in the menu for more details.
2023-03-01T04:57:20.810Z [info] : child process exited with code null
```

[specterApp copy pre33  Spectrum setup.log](https://github.com/cryptoadvance/specter-desktop/files/10877978/specterApp.copy.pre33.Spectrum.setup.log)

**Comment by k9ert** [comment]
The last lines in the logs seem to suggest that even though we have lots of "exit-messages" to be expected, none of those got executed. So it seems that the process got `killed -9` ?!
So, on the other hand, the specterd binary might be a bit alien to the system. Already since some time, unsigned binaries have [a hard time](https://github.molgen.mpg.de/pages/bs/macOSnotes/mac/mac_procs_unsigned.html) on a Mac. So manually executing specterd is probably worth a try. Is it even possible on a recvent Mac? 
As an other example, the bitcoind needs to be signed as explained [here](https://bitcoin.stackexchange.com/questions/117098/bitcoind-wont-run-on-macos).

So the suspicion is, that specterd is only starting up because of the way it got executed (not via a user-operated shell). And maybe there is a "cleaning mechanism" which shuts down "unauthorized" processes after returning from hibernation.

So, actions:
* Please try to execute `specterd` on the commandline checking if it is possible. 
<details>ysummary>on a older Mac it is still possible:</summary>
```
thomasmuller @ Thomass-Air.fritz.box ➜  ~  cd .specter/specterd-binaries
thomasmuller @ Thomass-Air.fritz.box ➜  specterd-binaries  ./specterd --help
Usage: specterd [OPTIONS] COMMAND [ARGS]...

Options:
  --debug                         Show debug information on errors.
  --tracerpc / --no-tracerpc      Will trace all calls to BitcoinCore or
                                  ElementsCore if in --debug
  --tracerequests / --no-tracerequests
                                  Will trace all calls done via the requests
                                  module. Might be quite verbose!
  --help                          Show this message and exit.

Commands:
  bitcoind   This will start a bitcoind regtest and mines a block every...
  elementsd  This will start a elements regtest and mines a block every...
  ext        Commands for the extension framework
  gunicorn   uses gunicorn instead of the build in development-server.
  server     This code is a function that runs Specter Desktop as a...
thomasmuller @ Thomass-Air.fritz.box ➜  specterd-binaries
```
</details>

* Check if signing of the specterd changes anything

**Comment by moritzwietersheim** [comment]
- Couldnt reproduce the issue this morning after over-night running Specter.
- But Specter was disconnected from Electrum this morning. Maybe bc of VPN.
- After reconnecting everything worked, and it synched with Electrum.
- I will keep it running and observe.

**Comment by moritzwietersheim** [comment]
- Running Specter overnight. It was disconnected from Electrum.
- Again, after reconnecting everything worked, and it synched with Electrum.

Logs:
[230305 1056 disconnected after overnight specterApp copy.log](https://github.com/cryptoadvance/specter-desktop/files/10892539/230305.1056.disconnected.after.overnight.specterApp.copy.log)
[230305 1104 reconnected specterApp copy.log](https://github.com/cryptoadvance/specter-desktop/files/10892540/230305.1104.reconnected.specterApp.copy.log)

- Had to close the MacBook for a few hours. And afterwards Specter was disconnected and need to be reconnected again. 

@k9ert, let's dig into this tomorrow my morning time.

**Comment by k9ert** [comment]
Here is a "remarkable line" filter on the latest log [230305 1104 reconnected specterApp copy.log](https://github.com/cryptoadvance/specter-desktop/files/10892540/230305.1104.reconnected.specterApp.copy.log)
```
SPECTERD: INFO in wallet_manager:   * failed_load_wallets: 0

2023-03-03T13:43:31.454Z [info] : stderr-SPECTERD: INFO in spectrum: process Notification: Electrum data {'jsonrpc': '2.0', 'method': 'blockchain.headers.subscribe', 'params': [{'hex': '00e01230bc962050918f7dd21b53b5c121226d980961fda936fa0000000000000000000069fe3633729904f3a356ec6d924bf6044d658fc1b1e47636965de76817a99d1e73f90164a3890617d2d71061', 'height': 779122}]
}
SPECTERD: INFO in spectrum: [{'hex': '00e01230bc962050918f7dd21b53b5c121226d980961fda936fa0000000000000000000069fe3633729904f3a356ec6d924bf6044d658fc1b1e47636965de76817a99d1e73f90164a3890617d2d71061', 'height': 779122}]

2023-03-03T13:46:39.700Z [info] : stderr-SPECTERD: INFO in spectrum: process Notification: Electrum data {'jsonrpc': '2.0', 'method': 'blockchain.headers.subscribe', 'params': [{'hex': '00e0ff3f40cc5dce3787a5734415accd53c9fe47109a10c33ed105000000000000000000d8988d6332d102c79ad648581ebefb77284bffd6884846f7fc3c26942865423a12fa0164a38906174d7cee2e', 'height': 779123}]
}
SPECTERD: INFO in spectrum: [{'hex': '00e0ff3f40cc5dce3787a5734415accd53c9fe47109a10c33ed105000000000000000000d8988d6332d102c79ad648581ebefb77284bffd6884846f7fc3c26942865423a12fa0164a38906174d7cee2e', 'height': 779123}]

2023-03-03T14:03:48.885Z [info] : stderr-SPECTERD: INFO in config_manager: Config loaded from file /Users/moritzwietersheim/.specter/config.json

2023-03-03T14:03:48.913Z [info] : stderr-SPECTERD: INFO in wallet_manager: Iterating over 2 wallet files in /Users/moritzwietersheim/.specter/wallets/main
SPECTERD: INFO in wallet_manager: Using threads in updating the wallet manager.
SPECTERD: INFO in wallet_manager: Started updating wallets with 2 wallets
SPECTERD: INFO in wallet_manager: Updating wallet MultiSig KTO

2023-03-03T14:03:50.036Z [info] : stderr-SPECTERD: INFO in wallet_manager: Updating wallet SingleSig Oven

2023-03-03T14:03:50.250Z [info] : stderr-SPECTERD: INFO in wallet_manager: Updating wallet manager done in  1352ms. Result:
SPECTERD: INFO in wallet_manager:   * loaded_wallets: 2
SPECTERD: INFO in wallet_manager:   * failed_load_wallets: 0

2023-03-03T14:04:31.777Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (1th time, next try in 10 seconds if threshold not met

2023-03-03T14:18:00.729Z [info] : stderr-SPECTERD: ERROR in requests_tools: SOCKSHTTPSConnectionPool(host='api.coindesk.com', port=443): Max retries exceeded with url: /v1/bpi/currentprice/USD.json (Caused by NewConnectionError('<urllib3.contrib.socks.SOCKSHTTPSConnection object at 0x1214980a0>: Failed to establish a new connection: [Errno 61] Connection refused'))
 while requesting https://api.coindesk.com/v1/bpi/currentprice/USD.json using proxies http socks5h://773098844:random@localhost:9050 and https socks5h://773098844:random@localhost:9050
SPECTERD: ERROR in requests_tools: There is a connection issue accessing the url https://api.coindesk.com/v1/bpi/currentprice/USD.json.
SPECTERD: ERROR in price_providers: There is a connection issue accessing the url https://api.coindesk.com/v1/bpi/currentprice/USD.json.. Check the logs for more details. while updating price

2023-03-03T14:18:01.072Z [info] : stderr-SPECTERD: INFO in config_manager: Config loaded from file /Users/moritzwietersheim/.specter/config.json
SPECTERD: INFO in wallet_manager: Iterating over 2 wallet files in /Users/moritzwietersheim/.specter/wallets/main
SPECTERD: INFO in wallet_manager: Using threads in updating the wallet manager.
SPECTERD: INFO in wallet_manager: Started updating wallets with 2 wallets
SPECTERD: INFO in wallet_manager: Updating wallet MultiSig KTO

2023-03-03T14:18:03.898Z [info] : stderr-SPECTERD: INFO in wallet_manager: Updating wallet SingleSig Oven

2023-03-03T14:18:04.622Z [info] : stderr-SPECTERD: INFO in wallet_manager: Updating wallet manager done in  3612ms. Result:
SPECTERD: INFO in wallet_manager:   * loaded_wallets: 2
SPECTERD: INFO in wallet_manager:   * failed_load_wallets: 0

2023-03-03T14:18:06.391Z [info] : stderr-SPECTERD: ERROR in elsock: Error in write: <class 'ssl.SSLEOFError'>

2023-03-03T14:18:46.390Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (2th time, next try in 10 seconds if threshold not met

2023-03-03T14:18:57.485Z [info] : stderr-SPECTERD: ERROR in elsock: Error in write: <class 'ssl.SSLEOFError'>

2023-03-03T14:19:36.409Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (3th time, next try in 10 seconds if threshold not met

2023-03-03T14:19:53.829Z [info] : stderr-SPECTERD: ERROR in elsock: Error in write: <class 'ssl.SSLEOFError'>

2023-03-03T14:20:31.640Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (4th time, next try in 10 seconds if threshold not met
SPECTERD: ERROR in elsock: More than 3 Ping failures for 30 seconds, Giving up!

2023-03-03T14:20:32.083Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from ok to broken_killing_threads
SPECTERD: INFO in elsock: Issue with Electrum deteted, threads died: ping
SPECTERD: INFO in elsock: trying to stop all threads ...

2023-03-03T14:20:32.084Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv write notify (0/2)

2023-03-03T14:20:32.101Z [info] : stderr-SPECTERD: INFO in elsock: Ended notify-loop

2023-03-03T14:20:32.889Z [info] : stderr-SPECTERD: INFO in elsock: Ended write-loop

2023-03-03T14:25:25.616Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (2/2)
[...]
2023-03-03T18:40:23.831Z [info] : stderr-SPECTERD: ERROR in requests_tools: SOCKSHTTPSConnectionPool(host='api.coindesk.com', port=443): Max retries exceeded with url: /v1/bpi/currentprice/USD.json (Caused by NewConnectionError('<urllib3.contrib.socks.SOCKSHTTPSConnection object at 0x12137a4a0>: Failed to establish a new connection: [Errno 61] Connection refused'))
 while requesting https://api.coindesk.com/v1/bpi/currentprice/USD.json using proxies http socks5h://1310881114:random@localhost:9050 and https socks5h://1310881114:random@localhost:9050
SPECTERD: ERROR in requests_tools: There is a connection issue accessing the url https://api.coindesk.com/v1/bpi/currentprice/USD.json.

2023-03-03T18:40:23.831Z [info] : stderr-SPECTERD: ERROR in price_providers: There is a connection issue accessing the url https://api.coindesk.com/v1/bpi/currentprice/USD.json.. Check the logs for more details. while updating price
```
So there was an internet connection issue at 13:43. The price-provider has difficulties to get to prices as well. All threads are exiting more or less immediately but not the recive thread which is still running hours later. At 20:15, we get requests again:
```
2023-03-03T20:07:58.881Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (424/106)

2023-03-03T20:15:21.204Z [info] : stderr-SPECTERD: INFO in config_manager: Config loaded from file /Users/moritzwietersheim/.specter/config.json

2023-03-03T20:15:21.224Z [info] : stderr-SPECTERD: INFO in wallet_manager: Iterating over 2 wallet files in /Users/moritzwietersheim/.specter/wallets/main

2023-03-03T20:15:21.225Z [info] : stderr-SPECTERD: INFO in wallet_manager: Using threads in updating the wallet manager.

2023-03-03T20:15:21.225Z [info] : stderr-SPECTERD: INFO in wallet_manager: Started updating wallets with 2 wallets

2023-03-03T20:15:21.227Z [info] : stderr-SPECTERD: INFO in wallet_manager: Updating wallet MultiSig KTO

2023-03-03T20:15:24.635Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (428/107)

2023-03-03T20:15:26.587Z [info] : stderr-SPECTERD: INFO in wallet_manager: Updating wallet SingleSig Oven

2023-03-03T20:15:26.749Z [info] : stderr-SPECTERD: INFO in controller: GET /settings/tor                            (200) took   10 ms

2023-03-03T20:15:26.813Z [info] : stderr-SPECTERD: INFO in controller: GET /wallets/wallets_loading/                (200) took    0 ms

2023-03-03T20:15:27.406Z [info] : stderr-SPECTERD: INFO in wallet_manager: Updating wallet manager done in  6180ms. Result:
SPECTERD: INFO in wallet_manager:   * loaded_wallets: 2
SPECTERD: INFO in wallet_manager:   * failed_load_wallets: 0

2023-03-03T20:15:27.830Z [info] : stderr-SPECTERD: INFO in controller: GET /wallets/wallets_loading/                (200) took    0 ms
```

Ok, so fast forward to the next day. Here, very similiar story. Connection was no longer there. Write loop has an ssl issue, ping died, monitoring stopped the rest of the threads, recv-thread doesn't want to exit.
```
2023-03-04T05:31:45.025Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (3th time, next try in 10 seconds if threshold not met

2023-03-04T05:31:45.740Z [info] : stderr-SPECTERD: ERROR in price_providers: Error as json doesn't look reasonable: 'http' while updating price

2023-03-04T05:31:55.137Z [info] : stderr-SPECTERD: ERROR in elsock: Error in write: <class 'ssl.SSLEOFError'>

2023-03-04T05:32:05.134Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (4th time, next try in 10 seconds if threshold not met
SPECTERD: ERROR in elsock: More than 3 Ping failures for 30 seconds, Giving up!

2023-03-04T05:32:05.892Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from ok to broken_killing_threads
SPECTERD: INFO in elsock: Issue with Electrum deteted, threads died: ping
SPECTERD: INFO in elsock: trying to stop all threads ...
SPECTERD: INFO in elsock: Waiting for those threads to exit: recv write notify (0/2)

2023-03-04T05:32:05.922Z [info] : stderr-SPECTERD: INFO in elsock: Ended notify-loop

2023-03-04T05:32:07.171Z [info] : stderr-SPECTERD: INFO in elsock: Ended write-loop

2023-03-04T05:32:15.915Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (2/2)

2023-03-04T05:32:25.935Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (4/2)
```
... still hours later, but then something interesting is happening. It seems like the user recreated the connection manually and even though the monitor-thread is still waiting on the recv-thread to exit, a new monitoring-thread got created ...
```
2023-03-04T13:26:19.993Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (340/85)

2023-03-04T13:26:40.014Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (344/86)

2023-03-04T13:26:41.130Z [info] : stderr-SPECTERD: INFO in service: Updating Spectrum node with electrum.emzy.de:50002 (ssl: True)
SPECTERD: INFO in spectrum: Stopping Spectrum

2023-03-04T13:26:41.130Z [info] : stderr-SPECTERD: INFO in spectrum: Creating ElectrumSocket electrum.emzy.de:50002 (ssl=True)
SPECTERD: INFO in elsock: Initializing ElectrumSocket with electrum.emzy.de:50002 (ssl: <module 'ssl' from '/var/folders/5m/jsvhmh7n661bxx9x1d4q6lc40000gn/T/_MEI0jh4MU/ssl.pyc'>) (proxy: None)

2023-03-04T13:26:41.130Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from unknown to creating_socket

2023-03-04T13:26:41.130Z [info] : stderr-SPECTERD: INFO in elsock: Started bg thread for _monitor_loop

2023-03-04T13:26:42.136Z [info] : stderr-SPECTERD: INFO in elsock: (re-)creating socket ...

2023-03-04T13:26:42.137Z [info] : stderr-SPECTERD: INFO in elsock: Connecting to electrum.emzy.de:50002

2023-03-04T13:26:43.511Z [info] : stderr-SPECTERD: INFO in elsock: Successfully created Socket <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.1.2.6', 54108), raddr=('135.181.215.237', 50002)> (ssl=True/tor=False)
SPECTERD: INFO in elsock: ElectrumSocket Status changed from creating_socket to creating_threads

2023-03-04T13:26:43.512Z [info] : stderr-SPECTERD: INFO in elsock: (re-)creating threads ...

2023-03-04T13:26:43.512Z [info] : stderr-SPECTERD: INFO in elsock: Started bg thread for recv_loop

2023-03-04T13:26:43.512Z [info] : stderr-SPECTERD: INFO in elsock: Started bg thread for _write_loop

2023-03-04T13:26:43.512Z [info] : stderr-SPECTERD: INFO in elsock: Started bg thread for _ping_loop

2023-03-04T13:26:43.512Z [info] : stderr-SPECTERD: INFO in elsock: Started bg thread for _notify_loop

2023-03-04T13:26:43.513Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from creating_threads to execute_recreation_callback
SPECTERD: INFO in elsock: ElectrumSocket Status changed from execute_recreation_callback to ok
2023-03-04T13:34:00.420Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (432/108)

2023-03-04T13:34:20.437Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (436/109)

2023-03-04T13:34:21.467Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (1th time, next try in 10 seconds if threshold not met

2023-03-04T13:34:40.462Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (440/110)

2023-03-04T13:34:41.481Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (2th time, next try in 10 seconds if threshold not met

2023-03-04T13:35:00.482Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (444/111)

2023-03-04T13:35:01.492Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (3th time, next try in 10 seconds if threshold not met

2023-03-04T13:35:20.500Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (448/112)

2023-03-04T13:35:21.500Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (4th time, next try in 10 seconds if threshold not met
SPECTERD: ERROR in elsock: More than 3 Ping failures for 30 seconds, Giving up!

2023-03-04T13:35:22.501Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from ok to broken_killing_threads
SPECTERD: INFO in elsock: Issue with Electrum deteted, threads died: ping

2023-03-04T13:35:22.501Z [info] : stderr-SPECTERD: INFO in elsock: trying to stop all threads ...
SPECTERD: INFO in elsock: Waiting for those threads to exit: recv write notify (0/2)

2023-03-04T13:35:22.507Z [info] : stderr-SPECTERD: INFO in elsock: Ended write-loop

2023-03-04T13:35:22.524Z [info] : stderr-SPECTERD: INFO in elsock: Ended notify-loop

2023-03-04T13:35:31.574Z [info] : stderr-SPECTERD: INFO in elsock: Ended recv-loop

2023-03-04T13:35:32.513Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from broken_killing_threads to creating_socket

2023-03-04T13:35:33.521Z [info] : stderr-SPECTERD: INFO in elsock: (re-)creating socket ...

2023-03-04T13:35:33.522Z [info] : stderr-SPECTERD: INFO in elsock: Connecting to electrum.emzy.de:50002

2023-03-04T13:35:35.390Z [info] : stderr-SPECTERD: INFO in elsock: Successfully created Socket <ssl.SSLSocket fd=6, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('192.168.1.122', 54424), raddr=('135.181.215.237', 50002)> (ssl=True/tor=False)

2023-03-04T13:35:35.391Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from creating_socket to creating_threads
SPECTERD: INFO in elsock: (re-)creating threads ...

2023-03-04T13:35:35.391Z [info] : stderr-SPECTERD: INFO in elsock: Started bg thread for recv_loop
SPECTERD: INFO in elsock: Started bg thread for _write_loop

2023-03-04T13:26:44.105Z [info] : stderr-SPECTERD: INFO in spectrum: Pinged electrum in 0.5089352130889893 
SPECTERD: INFO in spectrum: subscribe to block headers
```
but later on, the one of the monitoring-threads is still waiting on the recv-thread:
```
2023-03-04T13:28:40.127Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (368/92)
```
So some minutes later, the second monitoring thread (which should not have created in the first place btw) detects a failure, and this time successfully recreates the connection:
```
2023-03-04T13:34:00.420Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (432/108)

2023-03-04T13:34:20.437Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (436/109)

2023-03-04T13:34:21.467Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (1th time, next try in 10 seconds if threshold not met

2023-03-04T13:34:40.462Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (440/110)

2023-03-04T13:34:41.481Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (2th time, next try in 10 seconds if threshold not met

2023-03-04T13:35:00.482Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (444/111)

2023-03-04T13:35:01.492Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (3th time, next try in 10 seconds if threshold not met

2023-03-04T13:35:20.500Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (448/112)

2023-03-04T13:35:21.500Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (4th time, next try in 10 seconds if threshold not met
SPECTERD: ERROR in elsock: More than 3 Ping failures for 30 seconds, Giving up!

2023-03-04T13:35:22.501Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from ok to broken_killing_threads
SPECTERD: INFO in elsock: Issue with Electrum deteted, threads died: ping

2023-03-04T13:35:22.501Z [info] : stderr-SPECTERD: INFO in elsock: trying to stop all threads ...
SPECTERD: INFO in elsock: Waiting for those threads to exit: recv write notify (0/2)

2023-03-04T13:35:22.507Z [info] : stderr-SPECTERD: INFO in elsock: Ended write-loop

2023-03-04T13:35:22.524Z [info] : stderr-SPECTERD: INFO in elsock: Ended notify-loop

2023-03-04T13:35:31.574Z [info] : stderr-SPECTERD: INFO in elsock: Ended recv-loop

2023-03-04T13:35:32.513Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from broken_killing_threads to creating_socket

2023-03-04T13:35:33.521Z [info] : stderr-SPECTERD: INFO in elsock: (re-)creating socket ...

2023-03-04T13:35:33.522Z [info] : stderr-SPECTERD: INFO in elsock: Connecting to electrum.emzy.de:50002

2023-03-04T13:35:35.390Z [info] : stderr-SPECTERD: INFO in elsock: Successfully created Socket <ssl.SSLSocket fd=6, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('192.168.1.122', 54424), raddr=('135.181.215.237', 50002)> (ssl=True/tor=False)

2023-03-04T13:35:35.391Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from creating_socket to creating_threads
SPECTERD: INFO in elsock: (re-)creating threads ...

2023-03-04T13:35:35.391Z [info] : stderr-SPECTERD: INFO in elsock: Started bg thread for recv_loop
SPECTERD: INFO in elsock: Started bg thread for _write_loop
```
So while the second one is so healthy that sync is in progress, the first one still is waiting on the recv-thread to exit (see the different iteration-numbers):
```
2023-03-04T13:41:40.820Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (524/131)

2023-03-04T13:42:00.836Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (528/132)

2023-03-04T13:42:02.220Z [info] : stderr-SPECTERD: INFO in spectrum: Syncprocess now subscribed to 1200 scripthashes (100%, 3.1088082901554404 scripts/s)

2023-03-04T13:42:02.520Z [info] : stderr-SPECTERD: INFO in spectrum: Syncprocess finished syncing 1200 scripts in 386 with 3.1088082901554404 scripts/s)

2023-03-04T13:42:20.860Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (532/133)

2023-03-04T13:42:40.880Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (536/134)

2023-03-04T13:43:00.892Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (540/135)
```
On 14:34, we again have a connection issue but this time the second monitor thread can't recover either and both threads are waiting on the recv-thread:
```
2023-03-04T14:34:18.027Z [info] : stderr-SPECTERD: ERROR in elsock: Timeout in ping-loop (4th time, next try in 10 seconds if threshold not met
SPECTERD: ERROR in elsock: More than 3 Ping failures for 30 seconds, Giving up!

2023-03-04T14:34:18.782Z [info] : stderr-SPECTERD: INFO in elsock: ElectrumSocket Status changed from ok to broken_killing_threads
SPECTERD: INFO in elsock: Issue with Electrum deteted, threads died: ping
SPECTERD: INFO in elsock: trying to stop all threads ...
SPECTERD: INFO in elsock: Waiting for those threads to exit: recv write notify (0/2)

2023-03-04T14:34:18.807Z [info] : stderr-SPECTERD: INFO in elsock: Ended notify-loop

2023-03-04T14:34:19.073Z [info] : stderr-SPECTERD: INFO in elsock: Ended write-loop

2023-03-04T14:34:28.798Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (2/2)

2023-03-04T14:34:35.449Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (704/176)
```
And that goes on and on. In the last three lines we seem to have 3 broken monitor-threads waiting for a recv-thread each where at the same time another connection is processing properly.
```
2023-03-05T14:03:48.918Z [info] : stderr-SPECTERD: INFO in spectrum: process Notification: Electrum data {'jsonrpc': '2.0', 'method': 'blockchain.headers.subscribe', 'params': [{'height': 779448, 'hex': '0020002075fac67a0ca38bd205d71c5ac7f4e5c545548c14c02c04000000000000000000802abe19864272c20b58dc9d70e06bb609a4896df9f7d46119dd5557b79f6d0a2aa10464a38906172f72f74e'}]}

2023-03-05T14:03:48.918Z [info] : stderr-SPECTERD: INFO in spectrum: [{'height': 779448, 'hex': '0020002075fac67a0ca38bd205d71c5ac7f4e5c545548c14c02c04000000000000000000802abe19864272c20b58dc9d70e06bb609a4896df9f7d46119dd5557b79f6d0a2aa10464a38906172f72f74e'}]

2023-03-05T14:03:53.084Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (3468/867)

2023-03-05T14:03:56.790Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (2768/692)

2023-03-05T14:03:57.812Z [info] : stderr-SPECTERD: INFO in elsock: Waiting for those threads to exit: recv (1092/273)

```
So the bottom line:
* the recv-thread is not exiting properly even though he does sometimes for an unknown reason
* If the electrum connection is manually recreated, another instance of the spectrum/elsock will get created. This is probably not a good idea.

It would be great if the recv-thread would exit properly but i have no clue why that isn't working properly so i guess the next best option is to prevent the infra from being recreated and instead have a timeout in the monitor-thread which just moves on when the recv-thread is not exiting properly.

**Comment by k9ert** [comment]
I hope and think that this is fixed with this: https://github.com/cryptoadvance/spectrum/commit/31a0bac8bcc4920e32f40eb2a9dbf62cf2fe8dd6#diff-c896f39a2286d99dc02000a1a82ec07ffd39045cbef3b38eec1db8448bffaea7R471-R472 
which will get merged into specter-desktop with #2284
